// santa-main-r/lib/lib/viewer/get-rjs-config.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = getViewerRjsConfig;

var _rjsConfigUtils = require('../common/rjs-config-utils');

var _joinURL = require('../common/joinURL');

var _joinURL2 = _interopRequireDefault(_joinURL);

function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}

var defaults = {
    'santa-core-utils': '1.307.0',
    'santa-skin-utils': '1.1.0',
    'image-client-api': '1.40.0',
    imageClientLib: '1.40.0',
    zepto: '1.2.0',
    xss: '0.2.12',
    'react-dom-factories': '1.0.2',
    lodash: '4.7.0',
    'pm-rpc': '1.0.7',
    hammerjs: '2.0.8',
    // jsVendor: {
    mobx: '3.3.3',
    'mobx-react': '4.2.2',
    tweenmax: '1.19.0',
    'fedops-logger': '3.0.5'
};

var wixUiSantaStaticOverrideExp = 'allowWixUiSantaStaticOverride';

/**
 * @param {Object.<string, string>} versions
 * @param {*} serviceTopology
 * @param {boolean} local
 * @returns {*}
 */
function getViewerRjsConfig(versions, serviceTopology, local, experiment) {
    var platformSdkRoot = (0, _rjsConfigUtils.scriptLocation)(serviceTopology, 'js-platform-editor-sdk');
    var platformAppConfigurationUrl = (0, _rjsConfigUtils.scriptLocation)(serviceTopology, 'js-platform-apps-configuration', 'https://static.parastorage.com/services/js-platform-apps-configuration/1.7.0');
    var promoteAnalyticsAdapterUrl = (0, _rjsConfigUtils.scriptLocation)(serviceTopology, 'promote-analytics-adapter');
    var santaSiteAuthUrl = (0, _rjsConfigUtils.scriptLocation)(serviceTopology, 'santa-site-auth-module');

    (0, _rjsConfigUtils.mergeInto)(versions, defaults);

    /**
     * @type {function(pkg:string, path:string)}
     */
    var _unpkg = _rjsConfigUtils.unpkgOrNode.bind(null, versions, local);
    /**
     * @type {function(pkg:string, min:string, debug:string)}
     */
    var _unpkgObj = _rjsConfigUtils.unpkgOrNodeObj.bind(null, versions, local);

    var promoteAnalyticsAdapterPath = local ? (0, _rjsConfigUtils.node)('promote-analytics-adapter', 'dist/statics/analytics-event-adapter.bundle.min') : (0, _joinURL2.default)(promoteAnalyticsAdapterUrl, 'analytics-event-adapter.bundle.min');
    var santaSiteAuthPath = local ? (0, _rjsConfigUtils.node)('santa-site-auth-module', 'dist/statics/santa-site-auth-module.bundle.min') : (0, _joinURL2.default)(santaSiteAuthUrl, 'santa-site-auth-module.bundle.min');

    var paths = {
        // -------------- RUNTIME --------------------------------------------------------------
        wixImmutable: _unpkg('santa-core-utils', 'dist/wixImmutable'),
        lodash: _unpkg('lodash', 'lodash.min'),

        vendor: _unpkgObj('santa-bundle', 'dist/bundle.min', 'dist/bundle'),

        react: _unpkgObj('react', 'umd/react.production.min', 'umd/react.development'),
        'reactDOM': _unpkgObj('react-dom', 'umd/react-dom.production.min', 'umd/react-dom.development'),
        reactDOMServer: _unpkgObj('react-dom', 'umd/react-dom-server.browser.production.min', 'umd/react-dom-server.browser.development'),
        reactTestUtils: _unpkg('react-dom', 'umd/react-dom-test-utils.development'),
        'prop-types': _unpkgObj('prop-types', 'prop-types.min', 'prop-types'),
        'create-react-class': _unpkgObj('create-react-class', 'create-react-class.min', 'create-react-class'),
        'react-dom-factories': _unpkg('react-dom-factories', 'index'),
        mobx: _unpkgObj('mobx', 'lib/mobx.umd.min', 'lib/mobx.umd'),
        'mobx-react': _unpkgObj('mobx-react', 'index.min', 'index'),

        zepto: _unpkgObj('zepto', 'dist/zepto.min', 'dist/zepto'),
        immutable: _unpkgObj('immutable', 'dist/immutable.min', 'dist/immutable'),
        TweenMax: _unpkgObj('gsap', 'src/minified/TweenMax.min', 'src/uncompressed/TweenMax'), //bundle
        TweenLite: _unpkgObj('gsap', 'src/minified/TweenMax.min', 'src/uncompressed/TweenMax'),
        TimelineMax: _unpkgObj('gsap', 'src/minified/TweenMax.min', 'src/uncompressed/TweenMax'),
        ScrollToPlugin: _unpkgObj('gsap', 'src/minified/plugins/ScrollToPlugin.min', 'src/uncompressed/plugins/ScrollToPlugin'), //bundle
        Draggable: _unpkgObj('gsap', 'src/minified/utils/Draggable.min', 'src/uncompressed/utils/Draggable'),
        DrawSVGPlugin: _unpkgObj('santa-external-modules', 'tweenmax-plugins/1.19.0/DrawSVGPlugin.min', 'tweenmax-plugins/1.19.0/DrawSVGPlugin'),
        MorphSVGPlugin: _unpkgObj('santa-external-modules', 'tweenmax-plugins/1.19.0/MorphSVGPlugin.min', 'tweenmax-plugins/1.19.0/MorphSVGPlugin'),
        color: _unpkg('santa-external-modules', 'color-convert/0.2.0/color.min'),
        speakingurl: _unpkg('santa-external-modules', 'speakingurl/7.0.0/speakingurl.min'),
        'date-fns': _unpkgObj('santa-external-modules', 'date-fns/1.3.0/date_fns.min', 'date-fns/1.3.0/date_fns'),
        xss: _unpkg('xss', 'dist/xss.min'), //shim
        immutableDiff: _unpkg('santa-external-modules', 'immutableDiff/immutablejsdiff.min'),
        pmrpc: _unpkg('pm-rpc', 'build/pm-rpc.min'),
        'ag-grid': _unpkgObj('ag-grid', 'dist/ag-grid.min', 'dist/ag-grid'),

        SoundManager: _unpkg('soundmanager2', 'script/soundmanager2-nodebug-jsmin'),
        hammer: _unpkg('hammerjs', 'hammer.min'),
        mousetrap: _unpkg('mousetrap', 'mousetrap.min'),
        swfobject: _unpkg('santa-external-modules', 'swfobject/2.3.20130521/swfobject.min'),
        ajv: _unpkgObj('ajv', 'dist/ajv.min', 'dist/ajv.bundle'),
        'fedops-logger': _unpkg('fedops-logger', 'dist/statics/fedops-logger-module.bundle.min'),
        'fast-json-stable-stringify': _unpkg('santa-external-modules', 'dist/fast-json-stable-stringify'),

        // ------------- internal -------------------------------------------------------------------
        platformEvents: _unpkgObj('platform-editor-sdk', 'lib/platformEvents.min', 'lib/platformEvents'),
        'document-services': _unpkg('document-services', 'dist/document-services'),
        'documentServices': _unpkg('document-services-implementation', 'dist/document-services-implementation'),
        'document-services-implementation': _unpkg('document-services-implementation', 'dist/document-services-implementation'),
        'document-services-schemas': _unpkg('document-services-implementation', 'dist/document-services-schemas'),
        platformUtils: _unpkg('santa-platform-utils', 'dist/platformUtils-bundle'),
        mobileCore: _unpkg('santa-mobile-core', 'dist/mobileCore'),
        'santa-shared-schemas': _unpkg('santa-shared-schemas', 'dist/santa-shared-schemas'),
        'santa-components': _unpkgObj('santa-components', 'dist/santa-components.prod', 'dist/santa-components.devel'),
        'santa-components-layout': _unpkgObj('santa-components', 'dist/santa-components-layout.prod', 'dist/santa-components-layout.devel'),
        'santa-components/popover': _unpkgObj('santa-components', 'dist/santa-component-popover.prod', 'dist/santa-component-popover.devel'),
        'santa-renderer': _unpkg('santa-renderer', 'dist/santa-renderer'),
        mobileLayoutUtils: _unpkg('santa-mobile-core', 'dist/mobileLayoutUtils'),
        'mesh-migrator': _unpkg('mesh-migrator', 'dist/mesh-migrator'),
        skinUtils: _unpkg('santa-skin-utils', 'dist/skin-utils'),
        'santa-core-utils': _unpkg('santa-core-utils', 'dist/coreUtils'),
        warmupUtilsLib: _unpkg('santa-core-utils', 'dist/warmupUtils'),
        tweenEngine: _unpkg('santa-core-utils', 'dist/tweenEngine'),
        'image-client-api': _unpkg('image-client-api', 'dist/imageClientApi'),
        imageClientSDK: _unpkg('image-client-api', 'dist/imageClientSDK'),
        'santa-site-auth-module': santaSiteAuthPath,
        wixDomSanitizer: _unpkg('wix-dom-sanitizer', 'dist/wix-dom-sanitizer'),
        'santa-data-fixer': _unpkg('santa-data-fixer', 'dist/santa-data-fixer'),
        'santa-animations': _unpkg('santa-animations', 'dist/santa-animations'),
        'host-platform-api': _unpkg('santa-host-platform-services', 'dist/host-platform-api'),
        'host-worker-init': _unpkg('santa-host-platform-services', 'dist/host-worker-init'),
        wixFullstoryLoader: _unpkg('wix-fullstory-loader', 'dist/statics/app.bundle'),
        'data-capsule': _unpkg('data-capsule', 'dist/statics/frame-listener.bundle.min'),
        coreMultilingual: _unpkg('santa-multilingual', 'dist/languages'),
        'hls-light': _unpkg('hls.js', 'dist/hls.light.min'), //lazy
        hls: _unpkg('hls.js', 'dist/hls.min'), //lazy
        'promote-analytics-adapter': promoteAnalyticsAdapterPath,
        'js-platform-apps-configuration-editor': (0, _joinURL2.default)(platformAppConfigurationUrl, 'platform-apps-editor.min'),
        wixUiRichTextArea: _unpkg('wix-ui-santa', 'dist/statics/RichTextArea.bundle.min'),
        wixUiRichTextBox: _unpkg('wix-ui-santa', 'dist/statics/RichTextBox.bundle.min'),
        wixUiToggleSwitch: _unpkg('wix-ui-santa', 'dist/statics/ToggleSwitch.bundle.min'),
        wixUiTimePicker: _unpkg('wix-ui-santa', 'dist/statics/TimePicker.bundle.min'),
        wixUiVideoPlayer: _unpkg('wix-ui-santa', 'dist/statics/VideoPlayer.bundle.min'),
        wixUiPagination: _unpkg('wix-ui-santa', 'dist/statics/Pagination.bundle.min'),
        'wix-json-schema-utils': _unpkg('wix-json-schema-utils', 'dist/wix-json-schema-utils'),
        wixUiRating: _unpkg('wix-ui-santa', 'dist/statics/Rating.bundle.min'),
        wixUiSlider: _unpkg('wix-ui-santa', 'dist/statics/Slider.bundle.min'),
        wixUiTags: _unpkg('wix-ui-santa', 'dist/statics/Tags.bundle.min'),
        wixUiCompsToPackages: _unpkg('wix-ui-santa', 'dist/statics/compsToPackages.bundle.min'),
        wixUiSsrViewerCompsService: (0, _rjsConfigUtils.node)('wix-ui-santa', 'dist/src/components-service/viewer-ssr'),
        'wix-ui-santa': (0, _rjsConfigUtils.isExperimentOpen)(experiment, wixUiSantaStaticOverrideExp) ? (0, _rjsConfigUtils.scriptLocation)(serviceTopology, 'wix-ui-santa') : _unpkg('wix-ui-santa', 'dist/statics'),
        'stylable-panel': _unpkg('@wixc3/stylable-panel', 'dist/stylable-panel'),
        'stylable-panel-drivers': _unpkg('@wixc3/stylable-panel-drivers', 'dist/stylable-panel-drivers'),
        'stylable-santa-flatten': _unpkg('stylable-santa-flatten', 'dist/main'),
        'wix-base-ui': _unpkg('wix-base-ui', 'dist/base-ui'),
        'io-client': _unpkg('socket.io-client', 'dist/socket.io', 'dist/socket.io.dev'),
        'editorSkinsData': 'dist/editorSkinsData',

        // ------------- TESTING -------------------------------------------------------------------
        fake: 'js/plugins/fake/src/main/fake',
        definition: 'js/plugins/definition/src/main/definition',
        ReactProxy: _unpkg('santa-external-modules', 'react-proxy/ReactProxy'),
        Squire: _unpkg('santa-external-modules', 'squire/Squire'),
        jasmine: _unpkg('santa-external-modules', 'jasmine/2.1.3/jasmine'),
        'jasmine-html': _unpkg('santa-external-modules', 'jasmine/2.1.3/jasmine-html'),
        'jasmine-boot': _unpkg('santa-external-modules', 'jasmine/2.1.3/jasmine-boot'),
        // ------------- DEVELOPMENT ---------------------------------------------------------------
        io: 'https://cdnjs.cloudflare.com/ajax/libs/socket.io/1.4.0/socket.io.min',
        patcher: (0, _rjsConfigUtils.node)('santa-utils', 'common/hot/patcher'),
        hot: (0, _rjsConfigUtils.node)('santa-utils', 'common/hot/listener')
    };

    if (platformSdkRoot) {
        paths.platformAPI = {
            min: (0, _joinURL2.default)(platformSdkRoot, 'lib', 'platform-api.min'),
            source: (0, _joinURL2.default)(platformSdkRoot, 'lib', 'platform-api')
        };
    }

    var wixCodePlatformBaseUrl = (0, _rjsConfigUtils.scriptLocation)(serviceTopology, 'wix-code-platform');
    if (wixCodePlatformBaseUrl) {
        var getPathToWixCodePlatformFile = function getPathToWixCodePlatformFile(filename) {
            return local ? (0, _rjsConfigUtils.node)('wix-code-platform', 'dist/' + filename) : (0, _joinURL2.default)(wixCodePlatformBaseUrl, filename);
        };
        paths['elementory-browser-support'] = getPathToWixCodePlatformFile('elementory-browser-support.min');
        paths['wix-data-schemas-creator'] = {
            min: getPathToWixCodePlatformFile('wix-data-schemas-creator.min'),
            source: getPathToWixCodePlatformFile('wix-data-schemas-creator')
        };
    }

    return {
        //By default load any module IDs from js/lib
        baseUrl: '/',
        //except, if the module ID starts with "app",
        //load it from the js/app directory. paths
        //config is relative to the baseUrl, and
        //never includes a ".js" extension since
        //the paths config could be for a directory.
        paths: paths,
        bundles: {
            vendor: ['prop-types', 'create-react-class', 'mobx']
        },
        map: {
            '*': {
                imageClientLib: 'image-client-api',
                'react-dom': 'reactDOM',
                coreUtilsLib: 'santa-core-utils'
            }
        },
        shim: {
            color: {
                exports: 'Color'
            },
            'jasmine-html': {
                deps: ['jasmine']
            },
            'jasmine-boot': {
                deps: ['jasmine', 'jasmine-html']
            },
            SoundManager: {
                exports: 'soundManager'
            },
            ReactProxy: {
                deps: ['react'],
                exports: 'ReactProxy'
            },
            xss: {
                exports: 'filterXSS'
            }
        },
        waitSeconds: 0
    };
}
