// wix-custom-elements/dist/cjs/elements/wixBgImage/wixBgImage.js
'use strict';

var _createClass = function() {
    function defineProperties(target, props) {
        for (var i = 0; i < props.length; i++) {
            var descriptor = props[i];
            descriptor.enumerable = descriptor.enumerable || false;
            descriptor.configurable = true;
            if ("value" in descriptor) descriptor.writable = true;
            Object.defineProperty(target, descriptor.key, descriptor);
        }
    }
    return function(Constructor, protoProps, staticProps) {
        if (protoProps) defineProperties(Constructor.prototype, protoProps);
        if (staticProps) defineProperties(Constructor, staticProps);
        return Constructor;
    };
}();

function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}

function _possibleConstructorReturn(self, call) {
    if (!self) {
        throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }
    return call && (typeof call === "object" || typeof call === "function") ? call : self;
}

function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
        throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
    }
    subClass.prototype = Object.create(superClass && superClass.prototype, {
        constructor: {
            value: subClass,
            enumerable: false,
            writable: true,
            configurable: true
        }
    });
    if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
}

var imageLayout = require('./bgImageLayout');

function wixBgImageWrapper(WixElement, services, environmentConsts) {
    var WixBgImage = function(_WixElement) {
        _inherits(WixBgImage, _WixElement);

        function WixBgImage() {
            _classCallCheck(this, WixBgImage);

            // eslint-disable-line no-useless-constructor
            return _possibleConstructorReturn(this, _WixElement.call(this));
        }

        WixBgImage.prototype.reLayout = function reLayout() {
            var domNodes = {};
            var measures = {};

            var imageId = this.getAttribute('id');
            var imageInfo = JSON.parse(this.dataset.tiledImageInfo);
            domNodes[imageId] = this;
            domNodes[imageInfo.containerId] = document.querySelector('#' + imageInfo.containerId);
            imageInfo.displayMode = imageInfo.imageData.displayMode;

            services.mutationService.measure(function() {
                imageLayout.measure(imageId, measures, domNodes, imageInfo.containerId);
            });

            services.mutationService.mutate(function() {
                imageLayout.patch(imageId, measures, domNodes, imageInfo, environmentConsts);
            });
        };

        WixBgImage.prototype.attributeChangedCallback = function attributeChangedCallback(name, oldValue) {
            if (oldValue) {
                this.reLayout();
            }
        };

        WixBgImage.prototype.disconnectedCallback = function disconnectedCallback() {
            _WixElement.prototype.disconnectedCallback.call(this);
        };

        _createClass(WixBgImage, null, [{
            key: 'observedAttributes',
            get: function get() {
                return ['data-tiled-image-info'];
            }
        }]);

        return WixBgImage;
    }(WixElement);

    return WixBgImage;
}

module.exports = wixBgImageWrapper;
