// wix-custom-elements/dist/cjs/elements/wixDropdownMenu/wixDropdownMenu.js
'use strict';

var _createClass = function() {
    function defineProperties(target, props) {
        for (var i = 0; i < props.length; i++) {
            var descriptor = props[i];
            descriptor.enumerable = descriptor.enumerable || false;
            descriptor.configurable = true;
            if ("value" in descriptor) descriptor.writable = true;
            Object.defineProperty(target, descriptor.key, descriptor);
        }
    }
    return function(Constructor, protoProps, staticProps) {
        if (protoProps) defineProperties(Constructor.prototype, protoProps);
        if (staticProps) defineProperties(Constructor, staticProps);
        return Constructor;
    };
}();

function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}

function _possibleConstructorReturn(self, call) {
    if (!self) {
        throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }
    return call && (typeof call === "object" || typeof call === "function") ? call : self;
}

function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
        throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
    }
    subClass.prototype = Object.create(superClass && superClass.prototype, {
        constructor: {
            value: subClass,
            enumerable: false,
            writable: true,
            configurable: true
        }
    });
    if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
}

var dropdownMenuLayout = require('./dropdownMenuLayout');

function wixDropdownMenuWrapper(WixElement, services) {
    var WixDropdownMenu = function(_WixElement) {
        _inherits(WixDropdownMenu, _WixElement);

        function WixDropdownMenu() {
            _classCallCheck(this, WixDropdownMenu);

            // eslint-disable-line no-useless-constructor
            return _possibleConstructorReturn(this, _WixElement.call(this));
        }

        WixDropdownMenu.prototype.reLayout = function reLayout() {
            var menuId = this.getAttribute('id');
            var measures = void 0,
                domNodes = void 0;

            services.mutationService.measure(function() {
                var measureResult = dropdownMenuLayout.measure(menuId);
                measures = measureResult.measures;
                domNodes = measureResult.domNodes;
            });

            services.mutationService.mutate(function() {
                dropdownMenuLayout.patch(menuId, measures, domNodes);
            });
        };

        WixDropdownMenu.prototype.attributeChangedCallback = function attributeChangedCallback() {
            this.reLayout();
        };

        WixDropdownMenu.prototype.disconnectedCallback = function disconnectedCallback() {
            _WixElement.prototype.disconnectedCallback.call(this);
        };

        _createClass(WixDropdownMenu, null, [{
            key: 'observedAttributes',
            get: function get() {
                return ['data-hovered-item'];
            }
        }]);

        return WixDropdownMenu;
    }(WixElement);

    return WixDropdownMenu;
}

module.exports = wixDropdownMenuWrapper;
