// wix-custom-elements/dist/cjs/elements/wixImage/wixImage.js
'use strict';

var _createClass = function() {
    function defineProperties(target, props) {
        for (var i = 0; i < props.length; i++) {
            var descriptor = props[i];
            descriptor.enumerable = descriptor.enumerable || false;
            descriptor.configurable = true;
            if ("value" in descriptor) descriptor.writable = true;
            Object.defineProperty(target, descriptor.key, descriptor);
        }
    }
    return function(Constructor, protoProps, staticProps) {
        if (protoProps) defineProperties(Constructor.prototype, protoProps);
        if (staticProps) defineProperties(Constructor, staticProps);
        return Constructor;
    };
}();

function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}

function _possibleConstructorReturn(self, call) {
    if (!self) {
        throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }
    return call && (typeof call === "object" || typeof call === "function") ? call : self;
}

function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
        throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
    }
    subClass.prototype = Object.create(superClass && superClass.prototype, {
        constructor: {
            value: subClass,
            enumerable: false,
            writable: true,
            configurable: true
        }
    });
    if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
}

var imageLayout = require('./imageLayout');
var ImageLoader = require('./imageLoader');

var _require = require('../../utils/imageUtils'),
    getImageSrc = _require.getImageSrc;

var TIMEOUT = 250;

function wixImageWrapper(WixElement, services, environmentConsts) {
    if (!services.imageLoader) {
        services.imageLoader = new ImageLoader(services.mutationService);
    }

    var WixImage = function(_WixElement) {
        _inherits(WixImage, _WixElement);

        function WixImage() {
            _classCallCheck(this, WixImage);

            var _this = _possibleConstructorReturn(this, _WixElement.call(this)); // eslint-disable-line no-useless-constructor


            _this.childListObserver = null;
            _this.timeoutId = null;
            return _this;
        }

        WixImage.prototype.reLayout = function reLayout() {
            var domNodes = {};
            var measures = {};

            var imageId = this.getAttribute('id');
            var imageInfo = JSON.parse(this.dataset.imageInfo);
            var isSvg = this.dataset.isSvg === 'true';

            domNodes[imageId] = this;
            domNodes[imageInfo.containerId] = document.querySelector('#' + imageInfo.containerId);
            domNodes.image = this.querySelector(isSvg ? 'image' : 'img');
            domNodes.svg = isSvg ? this.querySelector('svg') : null;
            domNodes.picture = this.querySelector('picture');

            if (!domNodes.image) {
                // missing children, can't layout, wait for children to be created first
                // if isSvg==false || we don't have SVG yet -> use `this`
                var target = isSvg ? domNodes.svg || this : this;
                this.observeChildren(target);
                return;
            }

            // clean up
            this.unobserveChildren();

            // from now on just observe changes to children of top level
            this.observeChildren(this);

            // perform the layout
            services.mutationService.measure(function() {
                imageLayout.measure(imageId, measures, domNodes, imageInfo.containerId, isSvg);
            });

            var patchImage = function patchImage(shouldLoadImage) {
                services.mutationService.mutate(function() {
                    imageLayout.patch(imageId, measures, domNodes, imageInfo.imageData, imageInfo.alignType, services, environmentConsts, shouldLoadImage);
                });
            };

            // if image has no src - load the image immediately, otherwise - debounce the reload
            if (!getImageSrc(domNodes.image, isSvg)) {
                patchImage(true);
            } else {
                this.debounceImageLoad(patchImage);
            }
        };

        /**
         * Debounce consecutive image loads
         *
         * @param patchImage - closure for patching the image
         */


        WixImage.prototype.debounceImageLoad = function debounceImageLoad(patchImage) {
            clearTimeout(this.timeoutId);

            this.timeoutId = setTimeout(function() {
                patchImage(true);
            }, TIMEOUT);

            patchImage(false);
        };

        /**
         * Observe DOM mutations to wait for addition of missing children
         *
         * @param {HTMLElement} el
         */


        WixImage.prototype.observeChildren = function observeChildren(el) {
            var _this2 = this;

            if (!this.childListObserver) {
                this.childListObserver = new MutationObserver(function() {
                    return _this2.reLayout();
                });
            }

            this.childListObserver.observe(el, {
                childList: true
            });
        };

        /**
         * Remove DOM MutationObserver if one was created
         */


        WixImage.prototype.unobserveChildren = function unobserveChildren() {
            if (this.childListObserver) {
                this.childListObserver.disconnect();
                this.childListObserver = null;
            }
        };

        WixImage.prototype.attributeChangedCallback = function attributeChangedCallback(name, oldValue) {
            if (oldValue) {
                this.reLayout();
            }
        };

        WixImage.prototype.disconnectedCallback = function disconnectedCallback() {
            _WixElement.prototype.disconnectedCallback.call(this);
            services.imageLoader.onImageDisconnected(this);

            this.unobserveChildren();
        };

        _createClass(WixImage, null, [{
            key: 'observedAttributes',
            get: function get() {
                return ['data-image-info'];
            }
        }]);

        return WixImage;
    }(WixElement);

    return WixImage;
}

module.exports = wixImageWrapper;
