// @wix/fedops-logger/dist/es/src/reporter/reporter.js
'use strict';

exports.__esModule = true;

var _logger = require('@wix/web-bi-logger/dist/src/logger');

var defaultLoggerFactory = _interopRequireWildcard(_logger);

var _conf = require('../conf');

function _interopRequireWildcard(obj) {
    if (obj && obj.__esModule) {
        return obj;
    } else {
        var newObj = {};
        if (obj != null) {
            for (var key in obj) {
                if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key];
            }
        }
        newObj.default = obj;
        return newObj;
    }
}

function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}

var Reporter = function() {
    function Reporter() {
        var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
            biLoggerFactory = _ref.biLoggerFactory,
            baseUrl = _ref.baseUrl,
            endpoint = _ref.endpoint;

        _classCallCheck(this, Reporter);

        this.factory = biLoggerFactory ? biLoggerFactory : defaultLoggerFactory.factory({
            host: baseUrl
        });

        this.performancePublisher = this.factory.logger({
            endpoint: _conf.PERFORMANCE_BI_ENPOINT
        });
        this.fedopsPublisher = this.factory.logger({
            endpoint: endpoint || _conf.FEDOPS_BI_ENDPOINT
        });
    }

    Reporter.prototype.report = function report(data, reporterEndpoint) {
        if (!data) {
            return null;
        }

        switch (reporterEndpoint) {
            case _conf.PERFORMANCE_BI_ENPOINT:
                return this.performancePublisher.log(data);
            default:
                return this.fedopsPublisher.log(data);
        }
    };

    return Reporter;
}();

exports.default = Reporter;
