// @wix/fedops-logger/dist/es/src/data-items/common/bi-app-context-item.js
'use strict';

exports.__esModule = true;

var _dataItem = require('./data-item');

var _dataItem2 = _interopRequireDefault(_dataItem);

function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}

function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}

function _possibleConstructorReturn(self, call) {
    if (!self) {
        throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }
    return call && (typeof call === "object" || typeof call === "function") ? call : self;
}

function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
        throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
    }
    subClass.prototype = Object.create(superClass && superClass.prototype, {
        constructor: {
            value: subClass,
            enumerable: false,
            writable: true,
            configurable: true
        }
    });
    if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
}

var BiAppContextDataItem = function(_DataItem) {
    _inherits(BiAppContextDataItem, _DataItem);

    function BiAppContextDataItem(_ref) {
        var appId = _ref.appId,
            widgetId = _ref.widgetId,
            isServerSide = _ref.isServerSide,
            widgetArray = _ref.widgetArray;

        _classCallCheck(this, BiAppContextDataItem);

        var _this = _possibleConstructorReturn(this, _DataItem.call(this));

        _this.data = _this._filterUndefined({
            appId: appId,
            widgetId: widgetId,
            isServerSide: isServerSide,
            widgetArray: widgetArray
        });
        return _this;
    }

    BiAppContextDataItem.prototype._filterUndefined = function _filterUndefined(contextData) {
        var definedValues = {};
        Object.keys(contextData).forEach(function(key) {
            if (contextData[key]) {
                definedValues[key] = contextData[key];
            }
        });
        return definedValues;
    };

    return BiAppContextDataItem;
}(_dataItem2.default);

exports.default = BiAppContextDataItem;
