// @wix/web-bi-logger/dist/src/lib/navigator.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.listen = listen;

var _env = require('./env');

function hook(obj, fnName, hook) {
    if (!obj || !obj[fnName]) {
        return;
    }

    var original = obj[fnName];
    obj[fnName] = function() {
        hook.apply(null, arguments);
        return original.apply(obj, arguments);
    };
}

function onLoad(cb) {
    setTimeout(function() {
        return (0, _env.requestWindow)(function(window) {
            return cb(window.location.href);
        });
    });
}

function onPushState(cb) {
    (0, _env.requestWindow)(function(_ref) {
        var history = _ref.history;
        return ['pushState', 'replaceState'].forEach(function(fnName) {
            hook(history, fnName, function(_, __, url) {
                return cb(url);
            });
        });
    });
}

function onPopState(cb) {
    (0, _env.requestWindow)(function(window) {
        return window.addEventListener('popstate', function() {
            return cb(window.location.href);
        });
    });
}

function listen(cb) {
    (0, _env.requestWindow)(function(window) {
        var fromUrl = window.document.referrer;

        [onLoad, onPushState, onPopState].forEach(function(event) {
            return event(function(toUrl) {
                cb(fromUrl, toUrl);
                fromUrl = toUrl;
            });
        });
    });
}
