// @wix/web-bi-logger/dist/src/lib/tracker.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.track = track;

var _utils = require('./utils');

var _navigator = require('./navigator');

var _env = require('./env');

var _browser = require('./browser');

var IS_TRACKING = '__isTrackingBiBrowsingSession__';
var ENDPOINT = 'p';
var SOURCE = 19;
var EVID = 3;

function shouldTrack() {
    return !(0, _env.isWebWorker)() && (0, _env.requestWindow)(function(window) {
        return !window[IS_TRACKING];
    });
}

function setIsTracking() {
    return (0, _env.requestWindow)(function(window) {
        return window[IS_TRACKING] = true;
    });
}

function getLogger(loggerFactory) {
    return loggerFactory({
        endpoint: ENDPOINT,
        useBeacon: true
    }).setDefaults({
        src: SOURCE,
        evid: EVID,
        vsi: (0, _utils.guid)()
    }).logger();
}

function trackBrowsingSession(loggerFactory) {
    var logger = getLogger(loggerFactory);
    var firstInSession = 1;

    (0, _navigator.listen)(function(fromUrl, toUrl) {
        var _requestWindow = (0, _env.requestWindow)(function(window) {
                return {
                    sr: (0, _browser.getDesktopSize)(window),
                    wr: (0, _browser.getWindowSize)(window)
                };
            }),
            sr = _requestWindow.sr,
            wr = _requestWindow.wr;

        logger.log({
            from: fromUrl,
            to: toUrl,
            fis: firstInSession,
            sr: sr,
            wr: wr
        });
        firstInSession = 0;
    });
}

/**
 * Sends a page view event on initial load and on every page switch
 * BI event definition: http://bo.wix.com/bi-catalog-webapp/#/sources/19/events/3
 */
function track(loggerFactory) {
    if (shouldTrack()) {
        setIsTracking();
        trackBrowsingSession(loggerFactory);
    }
}
